### PR TITLE
feat(linters): add second guardrail wave

### PR DESCRIPTION
**Pull request by jonathanhaaswriter**
## Summary

This is the next stacked PR on top of #436. It lands the second guardrail wave before new rewrite code starts coming back into the tree.

### Added analyzers
- `nobackpointer` — forbid struct fields that point back to `*App` or `*Server`
- `sealedinterface` — forbid implementing `//cerebro:sealed` interfaces outside their home package
- `nopanicprod` — forbid `panic` outside tests, `init`, and `panicsafe`
- `noinmemorydb` — forbid embedded/in-memory database usage in production code
- `noenvoutsidecmd` — forbid `os.Getenv` / `os.LookupEnv` outside `cmd/` and `config`
- `nobackgroundctx` — forbid `context.Background` / `context.TODO` outside `cmd/` and tests

### Added arch check
- `tools/archtests/hooks_integrity_test.go`
- `tools/hooks/integrity.sha256`

This gives us a tracked hash manifest for `.githooks/pre-commit` and `.githooks/pre-push`, and wires a hook-integrity check into `make verify`.

### Build wiring
- register the new analyzers in `tools/linters/cerebrolint`
- add `check-arch` / `check-hook-integrity` targets
- make `check` and `verify` run the arch check

## Validation
- `make check-structural-test`
- `make check-arch`
- `make verify`

## Stack
- base: #436 (`feat/cerebro-next-remove-legacy-cerebro-20260422`)
- this PR: second structural guardrail wave before the rewrite primitives/store skeletons land

### CODE REVIEW

**Comment by jonathanhaaswriter** [comment]
@droid Please perform a deep code review and security review of this PR. Leave inline comments only for high-confidence, actionable issues.

**Comment by jonathanhaaswriter** [comment]
@droid Please re-review this PR after the latest fixes. Focus on whether the prior code review and security findings are resolved, and call out any remaining high-confidence issues.

**Comment by jonathanhaaswriter** [comment]
@droid Please re-review after the latest pushed fixes. Focus on whether the prior findings on this PR are resolved and only report remaining high-confidence issues.

**Comment by jonathanhaaswriter** [comment]
@droid Please re-review after the latest pushed fixes. Focus on whether the prior findings on this PR are resolved and only report remaining high-confidence issues.

**Comment by jonathanhaaswriter** [comment]
@droid Please re-review after the latest pushed fixes. Focus on whether the prior findings on this PR are resolved and only report remaining high-confidence issues.

**Comment by jonathanhaaswriter** [comment]
@droid please re-review

**Comment by jonathanhaaswriter** [comment]
@droid Please perform a fresh deep code review and security review of this PR at the latest head. Focus on whether the prior nopanicprod closure finding is resolved, and only leave inline comments for remaining high-confidence, actionable issues.

**Comment by jonathanhaaswriter** [comment]
@droid Please re-review after the latest pushed fixes. Focus on whether the log.Panic* and compile-time constant database guardrail findings are resolved, and only report remaining high-confidence actionable issues.

**Comment by jonathanhaaswriter** [comment]
@droid Please re-review after the latest pushed fixes. Focus on whether the dot-imported database/sql Open guardrail finding is resolved, and only report remaining high-confidence actionable issues.

File: Makefile [hunk 1]
```diff
@@ -31,7 +31,7 @@ hooks:
 pre-commit:
 	./scripts/pre_commit_checks.sh
 
-check: build test lint check-structural check-structural-test
+check: build test lint check-structural check-structural-test check-arch
 
 check-structural: check-structural-build
 	@$(LINTER_BIN) ./cmd/...
```
Comment (josh-writer) [inline Makefile]: We should probably have this also scan internal/ (etc) also so the new analyzers registered in tools/linters/cerebrolint/main.go can scan where they will be most useful

File: tools/linters/nobackpointer/nobackpointer.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+					Message: "struct field " + fieldLabel(field) + " stores back-pointer to *" + target + "; inject a narrow interface instead. (see PLAN.md §7 sin #7)",
+				})
+			}
+		}
+	})
+	return nil, nil
+}
+
+func backPointerTarget(pass *analysis.Pass, t types.Type) (string, bool) {
+	ptr, ok := t.(*types.Pointer)
```
Comment (josh-writer) [inline tools/linters/nobackpointer/nobackpointer.go]: backPointerTarget only handles direct *types.Pointer, 
so type AppRef *App followed by struct { app AppRef } bypasses the back-pointer rule while still storing a pointer to App.

File: tools/linters/sealedinterface/sealedinterface.go [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+	for obj := range sealed {
+		sealedObjects = append(sealedObjects, obj)
+	}
+	sort.Slice(sealedObjects, func(i int, j int) bool {
+		left := qualifiedTypeName(sealedObjects[i])
+		right := qualifiedTypeName(sealedObjects[j])
+		return left < right
+	})
+
+	for _, file := range pass.Files {
```
Comment (josh-writer) [inline tools/linters/sealedinterface/sealedinterface.go:90]: sealedinterface only checks local type declarations against sealed facts from imports; an external package can define a structurally matching type without importing the sealed package, and a third package can assign it to the sealed interface with no diagnostic.

File: tools/linters/nopanicprod/nopanicprod.go [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+			})
+			return true
+		})
+	}
+	return nil, nil
+}
+
+func isPanicCall(pass *analysis.Pass, call *ast.CallExpr) bool {
+	switch fun := call.Fun.(type) {
+	case *ast.Ident:
```
Comment (josh-writer) [inline tools/linters/nopanicprod/nopanicprod.go:144]: Dot-imported log panics bypass detection: Panic("boom") is an *ast.Ident, but the ident branch only checks builtin lowercase panic.

File: tools/linters/noenvoutsidecmd/noenvoutsidecmd.go [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+			pkgName, ok := pass.TypesInfo.Uses[pkgIdent].(*types.PkgName)
+			if !ok || pkgName.Imported() == nil || pkgName.Imported().Path() != "os" {
+				return
+			}
+			switch fun.Sel.Name {
+			case "Getenv", "LookupEnv":
+				report(fun.Sel.Name)
+			}
+		case *ast.Ident:
+			fn, ok := pass.TypesInfo.Uses[fun].(*types.Func)
```
[the remark this answers is not in the document]
Comment (josh-writer) [reply tools/linters/noenvoutsidecmd/noenvoutsidecmd.go]: nice catch

File: tools/linters/nobackgroundctx/nobackgroundctx.go [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+			pkgName, ok := pass.TypesInfo.Uses[pkgIdent].(*types.PkgName)
+			if !ok || pkgName.Imported() == nil || pkgName.Imported().Path() != "context" {
+				return
+			}
+			switch fun.Sel.Name {
+			case "Background", "TODO":
+				report(fun.Sel.Name)
+			}
+		case *ast.Ident:
+			fn, ok := pass.TypesInfo.Uses[fun].(*types.Func)
```
[the remark this answers is not in the document]
Comment (josh-writer) [reply tools/linters/nobackgroundctx/nobackgroundctx.go]: good catch

File: tools/linters/noenvoutsidecmd/noenvoutsidecmd.go [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+	return nil, nil
+}
+
+func fileAllowed(pass *analysis.Pass, pos token.Pos) bool {
+	return isTestFile(pass, pos) || isCmdFile(pass, pos) || isConfigPackage(pass)
+}
+
+func isCmdFile(pass *analysis.Pass, pos token.Pos) bool {
+	name := filepath.ToSlash(pass.Fset.Position(pos).Filename)
+	return strings.Contains(name, "/cmd/") || strings.HasPrefix(name, "cmd/")
```
Comment (josh-writer) [inline tools/linters/noenvoutsidecmd/noenvoutsidecmd.go]: NIT: isCmdFile checks the full checkout path, so a repo under a parent directory named cmd causes all files to be treated as allowed; the same helper in nobackgroundctx.go:75 has the same false-negative trigger. (unlikely edge case)

**Review by josh-writer — CHANGES_REQUESTED** [review]

**Comment by jonathanhaaswriter** [comment]
### Droid sweep status

Latest sweep: `2026-04-29 22:09 UTC`.

- Unresolved non-outdated review threads: `0`
- CI `verify`: `SUCCESS`
- `droid-review`: `not present`

I’ll keep watching for new review findings/check failures and address them as they appear.

**Comment by jonathanhaaswriter** [comment]
Re-audited all 11 live unresolved review threads (including outdated ones) against the current branch tip. Most feedback was already addressed on-branch; I made one follow-up fix for `nobackpointer` so it now catches back-pointers hidden inside container types/aliases, added regression coverage in `tools/linters/nobackpointer/testdata/src/a/a.go`, and pushed `c90fcc0f6` to this PR branch.

Validation run:
- `cd tools/linters && GOFLAGS= go test ./nobackpointer`
- `make verify` (also re-run by hooks on commit/push)

I inspected the staged diff before commit and did not stage any secrets. Remaining unresolved threads after this audit: 0.

**Comment by jonathanhaaswriter** [comment]
Droid sweep audit complete (final pass)

- Re-scanned all live review threads, including outdated ones
- Verified each remaining thread against the current branch tip
- Resolved every actionable thread (or fixed it inline and pushed)
- 0 unresolved review threads across the whole open-PR stack at this time

PR #494 (fix/feedback-main-followup) and PR #495 (feat/droid-creation-workflows) received additional fixes for SSRF (DNS-rebind) and workflow branch-name validation in this pass.

**Comment by jonathanhaaswriter** [comment]
@droid review

**Comment by jonathanhaaswriter** [comment]
@droid please harden this PR end-to-end:

1. Re-review the diff for residual bugs, races, edge cases, security gaps, and untested paths.
2. Apply the smallest correct fixes for any high-confidence issues you find.
3. Strengthen test coverage: add regression tests for every prior review finding on this PR plus the new behavior, including failure paths, boundary inputs, concurrency, and not-found / typed-nil interface guards.
4. Run focused tests and `make verify`. If validators fail, fix until green.
5. Commit and push the fixes to this PR's branch (or, if you lack push permission in this workflow, leave concrete inline suggestions and a summary comment).

Do not change the PR scope, do not modify the default branch, and do not merge.

**Comment by jonathanhaaswriter** [comment]
@droid review

File: tools/linters/sealedinterface/sealedinterface.go [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+func implementsSealed(named *types.Named, iface *types.Interface) bool {
+	return types.Implements(named, iface) || types.Implements(types.NewPointer(named), iface)
+}
+
+func reportImportedImplementation(pass *analysis.Pass, sealed map[*types.TypeName]*types.Interface, sealedObjects []*types.TypeName, targetType types.Type, value ast.Expr) {
+	sealedObj, iface := sealedTarget(sealed, sealedObjects, targetType)
+	if sealedObj == nil || iface == nil {
+		return
+	}
+	named := assignedNamedType(pass.TypesInfo.TypeOf(value))
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply tools/linters/sealedinterface/sealedinterface.go:195]: Fixed in 4fc2affb6 by unwrapping explicit interface conversions before inferring the concrete implementing type, with a regression for `sealedpkg.Runner(externalbad.Bad{})`.

File: tools/linters/noinmemorydb/noinmemorydb.go [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+		}
+	}
+
+	ins := pass.ResultOf[inspect.Analyzer].(*inspector.Inspector)
+	ins.Preorder([]ast.Node{(*ast.CallExpr)(nil)}, func(n ast.Node) {
+		call := n.(*ast.CallExpr)
+		if isTestFile(pass, call.Pos()) {
+			return
+		}
+		if ident, ok := call.Fun.(*ast.Ident); ok {
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply tools/linters/noinmemorydb/noinmemorydb.go:59]: Fixed in 4fc2affb6 by collecting aliases of `database/sql.Open`/`OpenDB` and diagnosing alias calls with sqlite/sqlite3 drivers, with local and package-level alias regressions.

File: tools/linters/nopanicprod/nopanicprod.go [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+		if fun.Name == "panic" {
+			return true
+		}
+		fn, ok := pass.TypesInfo.Uses[fun].(*types.Func)
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply tools/linters/nopanicprod/nopanicprod.go:148]: Fixed in 4fc2affb6 by collecting aliases of `log.Panic`/`Panicf`/`Panicln` and diagnosing alias calls, with local and package-level alias regressions.

**Comment by jonathanhaaswriter** [comment]
@droid review

Please re-review the guardrail linter wave deeply. I want adversarial coverage, not just happy-path analyzer tests:

1. **Alias/evasion resistance**: test import aliases, selector aliases, wrapper funcs, variable-assigned funcs, explicit interface conversions, dot imports, generated code, and package-local shadowing.
2. **False positives/negatives**: verify sealed-interface, no-in-memory-db, no-panic-prod, and related analyzers do not flag safe test-only code while still catching production bypasses.
3. **Type-system edge cases**: check generics, embedded interfaces, type aliases, pointer/value method sets, interface conversions in composite literals/returns, and multi-module package paths.
4. **Analyzer robustness**: ensure nil `types.Info`, parse/type errors, incomplete packages, and unusual AST forms do not panic or silently skip important files.
5. **CI/devex integration**: confirm the linters run in the intended package scope, produce actionable diagnostics, and do not regress performance enough to make hooks flaky.
6. **Regression tests**: look for missing testdata cases that would let an engineer bypass the guardrail with one small syntactic rewrite.

**Comment by jonathanhaaswriter** [comment]
Closing this draft PR — the underlying `feat(linters): add second guardrail wave` feature already landed on `main` via the earlier consolidation rollup (commit `3829c8cd8`), with `fix(linters): close guardrail bypasses` (`ee667162d`) following.

The additional bypass closures collected on this branch (e.g. `collectSQLOpenAliases`, `isEmbeddedDBPackageSelector`, dot-imported `sql.Open` detection, indirect panic closures, container-type back-pointers) represent a *new* layer of guardrail strengthening on top of what is already merged. Because this branch has diverged 80+ files from `main`, rebasing cleanly is impractical. If we still want those refinements, please open a fresh PR cut from `main` rather than reviving this one.
